### PR TITLE
Propagating the exceptions to the clients.

### PR DESCRIPTION
**Pull request by ankitaj224**

### CODE REVIEW

File: firebase-installations/src/main/java/com/google/firebase/installations/GetAuthTokenListener.java [hunk 1]
```diff
@@ -38,12 +38,13 @@ public boolean onStateReached(PersistedFidEntry persistedFidEntry) {
               .build());
       return true;
     }
+    return false;
+  }
 
+  @Override
+  public boolean onException(PersistedFidEntry persistedFidEntry, Exception exception) {
     if (persistedFidEntry.isErrored()) {
```
Comment (ciarand) [inline firebase-installations/src/main/java/com/google/firebase/installations/GetAuthTokenListener.java]: when would it be the case that we'd get an exception without the persistedFidEntry being in the error state?

File: firebase-installations/src/main/java/com/google/firebase/installations/GetIdListener.java [hunk 2]
```diff
@@ -27,15 +27,16 @@ public GetIdListener(TaskCompletionSource<String> taskCompletionSource) {
...
+  }
 
+  @Override
+  public boolean onException(PersistedFidEntry persistedFidEntry, Exception exception) {
     if (persistedFidEntry.isErrored()) {
-      taskCompletionSource.setException(
-          new FirebaseInstallationsException(
-              "Failed to update client side cache.",
-              FirebaseInstallationsException.Status.CLIENT_ERROR));
+      taskCompletionSource.trySetException(exception);
```
Comment (ciarand) [inline firebase-installations/src/main/java/com/google/firebase/installations/GetIdListener.java]: given this is the same in each of the listener impls so far we could move it into a default interface method

File: firebase-installations/src/main/java/com/google/firebase/installations/GetAuthTokenListener.java (hunk 1 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/GetAuthTokenListener.java]: Interrupted Exception 
Exception thrown while persisting. 
Thread failure

File: firebase-installations/src/main/java/com/google/firebase/installations/GetIdListener.java (hunk 2 above)
Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/GetIdListener.java]: error: [NoStaticOrDefaultMethodsInInterfaces] Avoid static/default methods in interfaces: We currently desugar SDKs with retrolambda which has limited support for those.

Default methods require API level 24. Current min is 14

Comment (ciarand) [reply firebase-installations/src/main/java/com/google/firebase/installations/GetIdListener.java]: oh, sad! I thought we had moved past retrolambda

File: firebase-installations/src/main/java/com/google/firebase/installations/StateListener.java [hunk 3]
```diff
@@ -18,4 +18,6 @@
 
 interface StateListener {
   boolean onStateReached(PersistedFidEntry persistedFidEntry);
+
+  boolean onException(PersistedFidEntry persistedFidEntry, Exception exception);
```
Comment (ciarand) [inline firebase-installations/src/main/java/com/google/firebase/installations/StateListener.java]: might be good to document the meaning behind the boolean return value

Comment (ankitaj224) [reply firebase-installations/src/main/java/com/google/firebase/installations/StateListener.java]: Updated. Thanks.

**Pull request merged**